### PR TITLE
Support requesting a specific IP from the plugin

### LINKED ISSUE #1734: Allow specification of specific IP addresses when using plugin
For users who want a static IP address.
Via the weave IPAM driver.
There is an `options` parameter passed in to the IPAM `RequestAddress` call, so presumably we can populate that somewhere in the Docker call.
Also will need to decide whether the request means "give me this address, even if you think it's already in use", "fail if I can't get this address" or "this is my preferred address but I'll take a different one if necessary".

### PR DESCRIPTION
**Pull request by bboreham**
Fixes #1734

### CODE REVIEW

**Comment by bboreham** [comment]
Rebased

File: plugin/ipam/driver.go [hunk 1]
```diff
@@ -81,11 +81,21 @@ func (i *ipam) RequestAddress(poolID string, address net.IP, options map[string]
...
 	}
-	// We are lying slightly to IPAM here: the range is not a subnet
-	if ip, err = i.weave.AllocateIPInSubnet("_", iprange); err != nil {
-		return
+	// See if we can give preferred address
+	if address != nil {
+		if err := i.weave.ClaimIP("_", address); err == nil {
+			ip = &net.IPNet{IP: address, Mask: subnet.Mask}
+		} else {
+			Log.Debugln("ClaimIP returned", err, "; allocating new address")
```
Comment (awh) [inline plugin/ipam/driver.go]: Is this the right thing to do, or should we fail the `RequestAddress` operation?

Comment (bboreham) [inline plugin/ipam/driver.go]: Well I thought it was, but looking at https://github.com/docker/libnetwork/blob/master/ipam/allocator.go#L492 it seems Docker's implementation does a fail.

It is documented as "preferred IP", so it seems strange to insist.

**Comment by bboreham** [comment]
I clarified the intended meaning, and removed the word "preferred" from Docker's code.

But there is a new problem: requesting the same address twice does not fail.

Currently we can't easily do a smoke-test because the `--ip` flag only comes in with Docker 1.10

**Comment by bboreham** [comment]
Fixed the problem that requesting the same address twice did not fail.

**Pull request merged**